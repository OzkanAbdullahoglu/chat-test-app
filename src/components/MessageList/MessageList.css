.list-root {
  list-style: none;
  box-sizing: border-box;
  box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.3);
  background-image: url('../../assets/images/background.jpg');
  padding-top: 8px;
  padding-bottom: 8px;
}

.main-list {
  width: 100%;
  margin: auto;
  max-width: 480px;
  overflow: auto;
  position: fixed;
  top: 120px;
  left: 0;
  right: 0;
  bottom: 60px;
}

.list-item-container {
  position: relative;
}

.main-list::-webkit-scrollbar {
  width: 1em;
}

.main-list::-webkit-scrollbar-track {
  box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.3);
}

.main-list::-webkit-scrollbar-thumb {
  background-color: #b6b3b3;
  outline: 1px solid rgb(153, 168, 184);
  border-radius: 1em;
}