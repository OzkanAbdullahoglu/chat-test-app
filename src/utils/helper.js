/* eslint-disable linebreak-style */
import React, { useState, useEffect } from 'react';
export const timeConverter = (timeStamp) => {
  if (!timeConverter.cache) {
    timeConverter.cache = {};
  }

  if (!timeConverter.cache[timeStamp]) {
    const timeStampMs = timeStamp * 1000;
    const dateObj = new Date(timeStampMs);
    const dayOfMonth = dateObj.getDate() < 10 ? `0${dateObj.getDate()}` : dateObj.getDate();
    const months = [
      'January', 'February', 'March', 'April', 'May', 'June',
      'July', 'August', 'September', 'October', 'November', 'December',
    ];
    const curMonth = months[dateObj.getMonth()];
    const curYear = dateObj.getFullYear();
    const curMinute = dateObj.getMinutes() < 10 ? `0${dateObj.getMinutes()}` : dateObj.getMinutes();
    const curHour = (dateObj.getHours() < 10 ? `0${dateObj.getHours()}` : dateObj.getHours());

    timeConverter.cache[timeStamp] = {
      date: `${curMonth} ${dayOfMonth},${curYear}`,
      time: `${curHour}:${curMinute}`,
    };
    return timeConverter.cache[timeStamp];
  }
  return timeConverter.cache[timeStamp];
};

export const isToday = (timestamp, date) =>{
  const today = Math.floor(Date.now() / 1000).toString();
  console.log(today);
  return timeConverter(today).date === timeConverter(timestamp).date ?
    'Today' : timeConverter(today).date === timeConverter(timestamp).date - 1 ?
    'Yesterday' : date;
};




function getCurrentViewDims() {
  const { innerWidth: width, innerHeight: height } = window;
  return {
    width,
    height,
  };
}

export default function useCurrentViewDims() {
  const [windowDimensions, setWindowDimensions] = useState(getCurrentViewDims());

  useEffect(() => {
    function handleResize() {
      setWindowDimensions(getCurrentViewDims());
    }

    window.addEventListener('scroll', handleResize);
    return () => window.removeEventListener('scroll', handleResize);
  }, []);

  return windowDimensions;
}

