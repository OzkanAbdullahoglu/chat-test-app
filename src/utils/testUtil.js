/* eslint-disable linebreak-style */
import { createStore, applyMiddleware } from 'redux';
import { middlewares } from '../store';
import rootReducer from '../reducers';

const chat = {
  requestedData: [
    {
      id: 1,
      direction: 'in',
      status: 'read',
      timestamp: '1577834102',
      text: 'Alice was beginning',
    },
  ],
  isPending: false,
  error: '',
  scrollDownEnable: true,
};

const message = {
  messagesEndRef: '',
};

const version = { number: '0.0.1' };

export const initialStates = { chat, message, version };

export const storeFactory = (initalState) => {
  const createStoreWithMiddleware = applyMiddleware(...middlewares)(createStore);
  return createStoreWithMiddleware(rootReducer, initalState);
};
