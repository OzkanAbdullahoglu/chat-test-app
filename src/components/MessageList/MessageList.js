/* eslint-disable linebreak-style */
import React, { useEffect } from 'react';
import { VariableSizeList } from 'react-window';
import { connect } from 'react-redux';
import { compose } from 'recompose';
import PropTypes from 'prop-types';

import { Wrapper } from '../Wrapper/Wrapper';
import { Icon } from '../Icon/Icon';
import TextInput from '../TextInput/TextInput';
import Row from '../Row/Row';
import './MessageList.css';
import {
  chatActions,
  getRequestedData,
  getShowTimeStampBool,
  getUnreadMessages,
  getScrollDownVisibilityStatus,
  getLastMessageRef,
} from '../../reducers';

const MessageList = ({
  updatedInitialData,
  setUpdatedTimeStampData,
  showTimeStampBool,
  unreadMesssages,
  setReadMessages,
  setToggleScrollDownDisable,
  setToggleScrollDownEnable,
  isScrollDownVisible,
  lastMessageInList,
  setDefault,
}) => {
  useEffect(() => { setUpdatedTimeStampData(); }, []);
  const listRef = React.createRef();
  const items = updatedInitialData;
  const defaultItemSize = 45;
  const doubleItemSize = 90;
  const getItemSize = (index) => showTimeStampBool((items[index].id)) ? doubleItemSize : defaultItemSize;
  const scrollToBottom = () => {
    if (lastMessageInList.current !== undefined) {
      lastMessageInList.current.scrollIntoView({
        behavior: 'smooth',
        block: 'end',
      });
    }
  };

  const handleClick = () => {
    const { firstUnreadId } = unreadMesssages;
    if (firstUnreadId !== 'none') {
      listRef.current.scrollToItem(firstUnreadId, {
        behavior: 'smooth',
        block: 'end',
      });
    }
  };

  const onItemsRendered = (item) => {
    const { firstUnreadId } = unreadMesssages;
    /* console.log(unreadMesssages);
    console.log(item, 'RENDER');*/
    const hideScrollButtonToBottom = 10;
    const visibleMessages = item.visibleStopIndex + 1;
    if (visibleMessages >= firstUnreadId) {
      setReadMessages(visibleMessages);
    }
    if (firstUnreadId === 'none' && items.length - visibleMessages < hideScrollButtonToBottom) {
      setToggleScrollDownDisable();
    } else {
      setToggleScrollDownEnable();
    }
  };

  return (
    <div className="main-list">
      <Wrapper
        incomingIcon
        scrollDown
        hideButton={!isScrollDownVisible}
      >
        <Icon
          incomingIcon
          scrollDown
          viewBox="0 0 48 48"
          onClick={handleClick}
        >
          <path xmlns="http://www.w3.org/2000/svg" d="M20 12l-2.83 2.83 9.17 9.17-9.17 9.17 2.83 2.83 12-12z" />
          <path xmlns="http://www.w3.org/2000/svg" d="M0 0h48v48h-48z" fill="none" />
        </Icon>
      </Wrapper>
      <VariableSizeList
        ref={listRef}
        height={615}
        width={480}
        itemSize={getItemSize}
        itemCount={items.length}
        className="list-root"
        overscanCount={30}
        onItemsRendered={onItemsRendered}
      >
        {Row}
      </VariableSizeList>
      <TextInput scrollToBottom={scrollToBottom} />
    </div>
  );
};

MessageList.propTypes = {
  updatedInitialData: PropTypes.array,
  showTimeStampBool: PropTypes.func,
  isScrollDownVisible: PropTypes.bool,
  setReadMessages: PropTypes.func,
  lastMessageInList: PropTypes.string,
  setUpdatedTimeStampData: PropTypes.func,
  setToggleScrollDownDisable: PropTypes.func,
  setToggleScrollDownEnable: PropTypes.func,
  unreadMesssages: PropTypes.object,
};

const mapStateToProps = (store) => ({
  updatedInitialData: getRequestedData(store),
  showTimeStampBool: getShowTimeStampBool(store),
  unreadMesssages: getUnreadMessages(store),
  isScrollDownVisible: getScrollDownVisibilityStatus(store),
  lastMessageInList: getLastMessageRef(store),
});

const withRedux = connect(
  mapStateToProps,
  { ...chatActions }
);


export default compose(withRedux)(MessageList);
