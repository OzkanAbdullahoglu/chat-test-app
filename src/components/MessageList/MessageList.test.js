/* eslint-disable linebreak-style */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import MessageList from './MessageList';
import { storeFactory } from '../../utils/testUtil';

const setup = (state = {}) => {
  const store = storeFactory(state);
  return shallow(<MessageList store={store} />).dive().dive();
};

describe('MessageList Component', () => {
  const messageList = setup();
  it('renders without crashing', () => {
    expect(toJson(messageList)).toMatchSnapshot();
  });
  it('contains Wrapper component', () => {
    expect(messageList.find('[data-test="wrapper"]').exists()).toBe(true);
  });
  it('contains SvgIcon component', () => {
    expect(messageList.find('[data-test="svg-icon"]').exists()).toBe(true);
  });
  it('contains Virtualised List component', () => {
    expect(messageList.find('[className="list-root"]').exists()).toBe(true);
  });
  it('contains TextInput component', () => {
    expect(messageList.find('Connect(TextInput)').exists()).toBe(true);
  });
});

describe('redux properties', () => {
  const store = storeFactory();
  const wrapper = shallow(<MessageList store={store} />).dive();
  it('has access to `requestedData` state', () => {
    const requestedData = [];
    const requestedDataProp = wrapper.props().store.getState().chat.requestedData;
    expect(requestedDataProp).toEqual(requestedData);
  });
  it('has access to `scrollDownEnable` state', () => {
    const isScrollDownEnable = true;
    const isScrollDownProp = wrapper.props().store.getState().chat.scrollDownEnable;
    expect(isScrollDownProp).toEqual(isScrollDownEnable);
  });
});
