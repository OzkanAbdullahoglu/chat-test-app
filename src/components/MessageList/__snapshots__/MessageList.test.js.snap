// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageList Component renders without crashing 1`] = `
<div
  className="main-list"
>
  <styled.div
    aria-label="Scroll down to the first unread message"
    data-test="wrapper"
    hideButton={false}
    incomingIcon={true}
    scrollDown={true}
    tabIndex="0"
  >
    <styled.svg
      data-test="svg-icon"
      incomingIcon={true}
      onClick={[Function]}
      scrollDown={true}
      viewBox="0 0 48 48"
    >
      <path
        d="M20 12l-2.83 2.83 9.17 9.17-9.17 9.17 2.83 2.83 12-12z"
        xmlns="http://www.w3.org/2000/svg"
      />
      <path
        d="M0 0h48v48h-48z"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
      />
    </styled.svg>
  </styled.div>
  <List
    className="list-root"
    direction="ltr"
    height={NaN}
    itemCount={0}
    itemSize={[Function]}
    layout="vertical"
    onItemsRendered={[Function]}
    overscanCount={5}
    useIsScrolling={false}
  >
    <Component />
  </List>
  <Connect(TextInput)
    scrollToBottom={[Function]}
    scrollToFirstUnread={[Function]}
  />
</div>
`;
